### DISCUSSION TITLE
Difficulties trying to use BF model

### DISCUSSION THREAD

**Discussion by kevinmuller98** [neuropoly/axondeepseg#751] [Help needed] [answered]
Hello everyone and greetings from Brazil!

I am conducting research on nerve regeneration after nerve injury and have just discovered AxonDeepSeg. I have no experience with Python/programming, but I successfully managed to install the program on Windows and run the testing models in napari (thanks for the very helpful guide on YouTube and the AxonDeepSeg website).

However, I am having some trouble trying to apply ADS in BF images. I am trying to learn how to use ADS with a test image of a Wistar rat sciatic nerve stained with osmium and toluidine blue, taken with a 100x objective lens (original image already in 8-bit).

![100X-5](https://github.com/axondeepseg/axondeepseg/assets/88518759/13a6c191-e54b-4131-8f2b-55c1b8d1d4a4)

but I don't know exactly how to set the ADS and running it with the default settings (pixel size in um = 0,070) is resulting me in the following image: 

![image](https://github.com/axondeepseg/axondeepseg/assets/88518759/74a64f47-15e3-428b-9bc9-24cd3b2a498d)

Using 0.0645 um/pixel, results in practically the same image above

But I am not sure about the right pixel-to-um conversion (100x objective lens = 0.0645 um/pixel?). Additionally, I am uncertain if I should do any type of treatment on the image before trying to run ADS. Do I need to do training with my images? Could you please advise me on what I might be doing wrong?

Thank you in advance for the development of the software and the amazing availability and support from the team. All the support and work on this forum are very thrilling!

Hugs, 
Kevin

**Comment by hermancollin** [comment, marked as answer]
Hi @kevinmuller98! Thank you for trying ADS and getting in touch.

> But I am not sure about the right pixel-to-um conversion (100x objective lens = 0.0645 um/pixel?). Additionally, I am uncertain if I should do any type of treatment on the image before trying to run ADS. Do I need to do training with my images? Could you please advise me on what I might be doing wrong?

Luckily, your image comes with a scale bar so we can figure this out easily. Using an image editing software (namely GIMP, but any other would work), I can see that the scale bar is 156 px wide. Since it represents 10 um, we get a pixel size of **0.0641 um/px**, so your assumption was correct. From a quick glance, I don't think you would need training for your images because the image quality is pretty good (although a little understained maybe?). Playing with the zoom factor helps here. The following segmentation was produced with a zoom factor of 0.75. This feature allows you to change the "effective" resolution of your image. For more information, see https://axondeepseg.readthedocs.io/en/latest/documentation.html#segment-images-using-a-zoom-factor

![Screenshot_20230727_140140](https://github.com/axondeepseg/axondeepseg/assets/83031821/23087008-e92d-4f94-bb61-05641a0341df)

However, as you can see, this segmentation is still unimpressive and misses most axons in the image. A known drawback of our default BF model is that it doesn't perform well when the axons have varying sizes. We do however have a better suited model for these kinds of samples. You can get it here: https://github.com/axondeepseg/model_seg_human_axon-myelin_bf/tree/master
To use it in `napari`, simply move the _model_seg_human_axon-myelin_bf_ folder at this location: `axondeepseg/AxonDeepSeg/models/` (where you installed ADS). See image below:
![Screenshot_20230727_141428](https://github.com/axondeepseg/axondeepseg/assets/83031821/c63270ac-a16c-41eb-a3fb-691b8564c5e8)

Using this model, I was able to obtain the following result (with zoom factor = 1.2), which now picks up most axons. Although it's not perfect, fine-tuning the zoom factor might help get a better segmentation.

![Screenshot_20230727_141544](https://github.com/axondeepseg/axondeepseg/assets/83031821/1d7d4c66-571d-4f77-ad37-8dc350a2fefb)

Any thoughts? Feel free to ask more questions and/or update us with your work.

Segmentations: [kevinmuller98_segmentations.zip](https://github.com/axondeepseg/axondeepseg/files/12186633/kevinmuller98_segmentations.zip)

**Reply by kevinmuller98** [reply]
Thank you for your answer, @hermancollin!!

I managed to use the human bf model, but I am still getting the same results:

![image](https://github.com/axondeepseg/axondeepseg/assets/88518759/f4202e2e-aba0-4e0d-a1d7-9e9b667d2104)

I tried to change the zoom factor through the settings box in napari, but it seems that it is not adjusting properly. I altered the value (0.7, 1, 1.5, 2, 2.75) but it keeps giving me the same axon measurements. Maybe I didn't selected something that is important?

Another question that I have is regarding the morphometrics result Excel. All the morphometrics values of each axon are presented in the same Excel cell. Is there a way that I can sort the measurements (area, perimeter, g ratio, etc.) into different cells? Or how could I arrange them to take a mean value of each parameter of all analyzed axons?

Thank you for all your help!!

**Reply by hermancollin** [reply]
@kevinmuller98 hmm this is not normal. Comparing with your original message, the segmentation is identical. Before clicking on `Apply ADS model`, did you delete the previous masks on the left panel? 

![Screenshot_20230728_125656](https://github.com/axondeepseg/axondeepseg/assets/83031821/1c252e69-000f-41e7-92f9-94be82fe4a72)

> Another question that I have is regarding the morphometrics result Excel. All the morphometrics values of each axon are presented in the same Excel cell. Is there a way that I can sort the measurements (area, perimeter, g ratio, etc.) into different cells? Or how could I arrange them to take a mean value of each parameter of all analyzed axons?

Can you send a screenshot? I'm not sure I understand. Maybe excel is not importing the file properly

**Reply by kevinmuller98** [reply]
I think I did delete the previously used masks as I was changing the zoom factor. But as I remade the analysis from 0 today, reopening the ADS and napari, it is strange to get the same results as yesterday. But, I tried it again now and it seems to be correctly working, maybe something did not load properly before:

![image](https://github.com/axondeepseg/axondeepseg/assets/88518759/53e8bf24-f796-43e5-ad32-9abb931bde15)

As I increase the zoom factor, got different result: 

![image](https://github.com/axondeepseg/axondeepseg/assets/88518759/e228107c-de28-4bf2-9540-94332aacb7fd)

Trying with different images, I did also got good results:

![image](https://github.com/axondeepseg/axondeepseg/assets/88518759/9edd739c-a537-4167-bba1-3be5658472a4)


Regarding the morphometric, this is how the data is showing: 

![image](https://github.com/axondeepseg/axondeepseg/assets/88518759/7a1a9481-586e-41be-912a-ce67389da5d9)

All the numbers should appear in the same cell separeted by ","?

Thanks, 

Kevin

**Reply by Stoyan-I-A** [reply]
Regarding the morphometrics file. On some systems, Excel will be configured to read `.csv` files using semi-colons (`;`) or tabs as the separating character. 
You can try to import the `.csv` file into Excel by opening a new sheet, then go to `Data -> Get Data -> From File -> From Text/CSV` and change the separating character to comma.

**Reply by kevinmuller98** [reply]
Thanks for the advice @Stoyan-I-A .

It did work to arrange the data into different cells, but it messed up the values.
![image](https://github.com/axondeepseg/axondeepseg/assets/88518759/fc6198b3-422f-49d7-b55f-61ff0d6d7736)

i.e., 
row 0 in the adjusted excel: 0 | 3,70467E+15 | 6,06226E+15 | 4,26628E+15 | 106918425 | 4,07725E+14 | 4,80509E+16 | 1,16676E+16 | 7,84041E+15 | 5874273 | 1,56442E+16 | 9,44853E+15 | 8,03172E+15 | -1,22285E+15

row 0 in the .csv:  0,37.04669260700389,6.062256809338521,0.4266278606054218,1.06918425,4.07725162159597,4.8050887499999995,1.1667594729454767,0.7840408901241092,5.874273,15.644165522408798,0.9448529411764706,0.8031724040842376,-1.222853703505916

Did I do something wrong?

**Reply by hermancollin** [reply]
@kevinmuller98 yes something went wrong. Excel read all the values as integers instead of floating points. 

You mentioned you were from Brazil. If I'm not mistaken, in Portugese the decimal separator is the comma, not the point. I actually didn't know that (we have the same in French). You should be able to change it by going in **File > Options**, then **Advanced > Editing options > Use system separator**. Uncheck this box and change the comma to a point. Maybe this will work.

**Reply by mathieuboudreau** [reply]
@hermancollin 

>However, as you can see, this segmentation is still unimpressive and misses most axons in the image. A known drawback of our default BF model is that it doesn't perform well when the axons have varying sizes. We do however have a better suited model for these kinds of samples. You can get it here: https://github.com/axondeepseg/model_seg_human_axon-myelin_bf/tree/master
To use it in napari, simply move the model_seg_human_axon-myelin_bf folder at this location: axondeepseg/AxonDeepSeg/models/ (where you installed ADS). See image below:

Is there any reason why we haven't included this model into the main software directly yet?

**Reply by hermancollin** [reply]
@mathieuboudreau it was suggested in #626 but apparently we chose not to add it during a meeting.

At this point I think we should re-open this issue. In the last months I explained to a few users how to manually use this model but it's not really practical. It appears this model works much better for certain BF images

**Comment by kevinmuller98** [comment]
> @kevinmuller98 yes something went wrong. Excel read all the values as integers instead of floating points.
> 
> You mentioned you were from Brazil. If I'm not mistaken, in Portugese the decimal separator is the comma, not the point. I actually didn't know that (we have the same in French). You should be able to change it by going in **File > Options**, then **Advanced > Editing options > Use system separator**. Uncheck this box and change the comma to a point. Maybe this will work.

Yes, here in Brazil, we also use the comma as a separator. I changed the system separator, but it's still making alterations in the values.

![image](https://github.com/axondeepseg/axondeepseg/assets/88518759/33a2effb-28f1-423e-9d38-3ec2f69435e4)

**Reply by Stoyan-I-A** [reply]
What happens if you select all the cells, then in the Home tab, you change the number format from `General` to `Text`?

**Reply by kevinmuller98** [reply]
@Stoyan-I-A nothing happens :/

I made some progress through the Data -> Text to Columns -> Delimited -> changed delimiters from 'tab' to 'comma' (selected only the first rows):

![image](https://github.com/axondeepseg/axondeepseg/assets/88518759/dc7e7bb3-9ae4-4814-9768-24e534589056)

It worked to separate the data into columns, but it also messed up some values, such as the axon_area or axon_perimeter. Then, I changed the system separator as suggested, and it worked!!!

![image](https://github.com/axondeepseg/axondeepseg/assets/88518759/00fe0f54-9dbe-4d57-a794-c8331149200e)

Do these values seem alright?

Thank you for all the help, guys!

**Reply by Stoyan-I-A** [reply]
They look okay in your last image

**Reply by kevinmuller98** [reply]
Thanks for the help, guys! I am really glad that it worked and anxious to put the ADS to work. Hope to cite you all in future publications! See ya!
